ext2/ext2.c

```diff
@@ -122,20 +129,18 @@ int ext2_lookup(ext2_t *fs, id_t id, const char *name, uint8_t len, oid_t *res,
 		mutexUnlock(dir->lock);
 		ext2_obj_put(fs, dir);
 
-		if (err < 0)
+		if (err < 0) {
 			return err;
-	}
-
-	mutexLock(obj->lock);
+		}
 
-	if (S_ISCHR(obj->inode->mode) || S_ISBLK(obj->inode->mode)) {
-		memcpy(dev, &obj->dev, sizeof(oid_t));
-	}
-	else {
-		dev->port = fs->port;
-		dev->id = res->id;
+		mutexLock(obj->lock);
```

Comment:
lock taken in `for`, released only after exiting from `for` - might be taken many times (i.e. 2 times, resulting in deadlock)